iree/compiler/Codegen/Common/FoldAffineMinInDistributedLoops.cpp

```diff
@@ -0,0 +1,240 @@
+/// distributed to processors, where we have the structure:
+///
+/// ```mlir
+/// %id = hal.interface.workgroup.id ...
+/// %count = hal.interface.workgroup.count ...
+/// %offset = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%id, ...]
+/// %size = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%count, ...]
+/// scf.for %iv = %offset to ... step %size { ... }
+/// ```
+///
+/// For such loops, we need to ignore the distribution aspect and get the
+/// lower bound, upper bound, and step for only the tiling aspect, so that we
+/// can reuse upstream utilities to prove that the `affine.min` ops are tightly
+/// bound so that we can replace them with the tight bound.
+struct FoldAffineMinOverDistributedLoopInductionVariable final
+    : public OpRewritePattern<AffineMinOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AffineMinOp minOp,
+                                PatternRewriter &rewriter) const override {
+    QueryProcessor queryProcessor{getProcessorIDDim, getProcessorCountDim,
+                                  getProcessorTileSizeDim};
+    Location loc = minOp.getLoc();
+
+    auto loopMatcher = [&](Value iv, Value &lb, Value &ub, Value &step) {
+      scf::ForOp forOp = scf::getForInductionVarOwner(iv);
+      if (!forOp) return failure();
+
+      auto loopInfo = isTiledAndDistributedLoop(forOp, queryProcessor);
+      if (!loopInfo) return failure();
+      LLVM_DEBUG(llvm::dbgs() << *loopInfo);
+
+      Optional<int64_t> untiledStep =
+          getConstantIntValue(loopInfo->untiledStep);
+      // For IREE right now the original untiled loop should have step 1..
+      if (!untiledStep || *untiledStep != 1) return failure();
+      // ..and we tile according to some static tile sizes for processors.
+      if (!loopInfo->tileSize) return failure();
+
+      lb = getAsValue(loopInfo->untiledLowerBound, rewriter, loc);
+      ub = getAsValue(loopInfo->untiledUpperBound, rewriter, loc);
+      // The "step" expected by the upstream utility is really the tiling size.
+      step = rewriter.create<arith::ConstantIndexOp>(
+          loc, loopInfo->tileSize.getValue());
+      return success();
+    };
+
+    return scf::canonicalizeMinMaxOpInLoop(
+        rewriter, minOp, minOp.getAffineMap(), minOp.operands(), /*isMin=*/true,
+        loopMatcher);
+  }
+};
+
+/// Folds `affine.apply` ops over induction variables that actually can only
+/// take one single value.
+//
+// TODO: This should be removed and use the above pattern. But for some reason
+// the above pattern does not handle a specific corner case:
+//
+// ```mlir
+// scf.for %iv = %c0 to %c4 step %c4 {
+//   %0 = affine.min affine_map<(d0, d1)[] -> (4, d1 - d0)> (%iv, %c4)
+// ```
+//
+// The above will be folded into
+//
+// ```mlir
+// scf.for %iv = %c0 to %c4 step %c4 {
+//   %0 = affine.apply affine_map<(d0) -> (-d0 + 4)>(%iv)
+// ```
+//
+// But we would expect `%0` to be `%c4` entirely. It looks like a bug. So use
```

Comment:
This definitely looks like a bug.

Reply:
Yeah. I'll chat with Matthias and give him a repro; he can probably look and fix it later.